File: 25. Reverse Nodes in k-Group/main.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

import (
	. "com.grpk.leet/utils/LinkedListUtils"
)

func main() {
	ans := reverseKGroup(GenFromSlice([]int{}), 3)
	PrintListNode(ans)
}

/**
* r: decide whether need return
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
	newHead := &ListNode{
		Next: head,
	}
	curr := newHead
	r := newHead
	for curr.Next != nil {
		for i := 0; i < k; i++ {
			if r.Next == nil {
				return newHead.Next
			}
			r = r.Next
		}
		next := r.Next

		nextHead := reverse(curr.Next, k-1)
		// curr: 0
		temp := curr.Next    // old head eg 1
		curr.Next = nextHead // update to new head eg 0->2->1
		curr = temp          // new curr: old head
		r = curr
		curr.Next = next // connect new tail(old head) to next part
	}
	return newHead.Next
}

func reverse(head *ListNode, count int) *ListNode {
	if head.Next == nil || count == 0 {
		return head
	}

	newHead := reverse(head.Next, count-1)
	head.Next.Next = head
	head.Next = nil

	return newHead
}

File: 25. Reverse Nodes in k-Group/main.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def reverse(head: ListNode, stop: ListNode) -> ListNode:
            curr = head
            prev = None
            while not curr == stop:
                next = curr.next
                curr.next = prev
                prev = curr
                curr = next
            
            return prev

        if head == None:
            return None

        start: ListNode = head
        end: ListNode = head
        for i in range(0, k):
            if not end:
                return head
            end = end.next
        
        newHead = reverse(start, end)
        start.next = self.reverseKGroup(end, k)

        return newHead

s = Solution()
s.reverseKGroup(ListNode(1,ListNode(2, ListNode(3, ListNode(4, ListNode(5))))), 2)